content/blog/2019-11-13-llvm-autovec/index.md

```diff
@@ -0,0 +1,213 @@
+  br i1 %cmp, label %for.body, label %for.end
+
+for.body:                                         ; preds = %for.cond
+  %arrayidx = getelementptr [i64]* %a, i64 0, i64 %i.0
+  store i64 %i.0, i64* %arrayidx, align 8
+  br label %for.inc
+
+for.inc:                                          ; preds = %for.body
+  %inc = add nsw i64 %i.0, 1
+  br label %for.cond
+
+for.end:                                          ; preds = %for.cond
+  ret i32 0
+```
+
+Running our vectorization pass, we get the following:
+
+```
+entry:
+  %a = alloca [100 x i64], align 16
+  br label %for.cond
+
+for.cond:                                         ; preds = %for.inc, %entry
+  %i.0 = phi i64 [ 0, %entry ], [ %inc, %for.inc ]
+  %0 = insertelement <4 x i64> zeroinitializer, i64 %i.0, i64 0
+  %1 = insertelement <4 x i64> %0, i64 %i.0, i64 1
+  %2 = insertelement <4 x i64> %1, i64 %i.0, i64 2
+  %3 = insertelement <4 x i64> %2, i64 %i.0, i64 3
+  %4 = add <4 x i64> %3, <i64 0, i64 1, i64 2, i64 3>
+  %cmp = icmp slt i64 %i.0, 100
+  br i1 %cmp, label %for.body, label %for.end
+
+for.body:                                         ; preds = %for.cond
+  %arrayidx = getelementptr [i64]* %a, i64 0, i64 %i.0
+  %5 = bitcast i64* %arrayidx to <4 x i64>*
+  store <4 x i64> %4, <4 x i64>* %5, align 8
+  br label %for.inc
+
+for.inc:                                          ; preds = %for.body
+  %inc = add nsw i64 %i.0, 4
+  br label %for.cond
+
+for.end:                                          ; preds = %for.cond
+  ret i32 0
+```
+
+The vectorizer changes the loop so that each new iteration performs 4 iterations
+of the original loop.
+The GEP in the `for.body` block computes the base address for a 4-vector;
+the subsequent `store` into `%arrayidx` then uses the vectorized version 
+of the inductive variable `%i.0` computed in `for.cond` (`%4`)
+to write into memory the next 4 values of `%i.0` at once.
+
+
+## Evaluation
+
+We evaluated our autovectorizer across a range of small benchmarks containing
+loops.
+We constructed our benchmarks so that all loops can be vectorized according
+to the rather stringent vectorization conditions, which disallows indexing
+arrays with anything other than the inductive variable or loop-invariant data.
+The results below are measured across three runs.
+Note that DEF is the configuration without any optimizations (`-O0`),
+while OPT is the configuration with only our autovectorizer enabled.
+
+Benchmark   | DEF runtime (ms) | OPT runtime (ms) 
+------------|------------------|------------------
+test1       | 26.00            | 13.00            
+test2       | 25.00            | 22.67            
+test3       | 19.00            | 15.33            
+test4       | 25.33            | 15.00            
+test5       | 10.00            | 11.00
+
+Across the benchmarks, either OPT outperforms DEF handily
+(as in `test1` and `test4`), or it runs about the same time
+as its DEF counterpart.
+This is the behavior we expected.
```

Comment:
It would be interesting to investigate a little deeper to determine why there was a difference in performance advantage. What is it about `test1` and `test4` that is different from the others? It's hard to tell without a description of the benchmarks.